enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,891 @@
+registry server, and without requiring OpenShift knowledge for the technicians
+performing the upgrades.
+
+## Motivation
+
+All these stories are in the context of disconnected clusters with limited
+resources, both in the cluster itself and in the surrounding environment:
+
+- The cluster is not connected to the Internet, or the bandwidth is very limited.
+- It isn't possible to bring up additional machines, even temporarily.
+- The resources of the cluster are limited, in particular in terms of CPU and memory.
+- The technicians responsible for performing the upgrade have little or no OpenShift knowledge.
+
+These clusters are usually installed at the customer site by a partner engineer
+collaborating with customer technicians.
+
+Eventually, the cluster will need to be upgraded, and then the technicians will
+need tools that make the process as simple as possible, ideally requiring no
+OpenShift knowledge.
+
+When OpenShift knowledge is required the technician performing the upgrade will
+have the support of the team of engineers that planned and vetted the upgrade.
+
+### User Stories
+
+#### Pre-load and pin upgrade images
+
+As an engineer managing a cluster that has a low bandwidth and/or unreliable
+connection to an image registry server I want to pin and pre-load all the
+images required for the upgrade so that when I decide to actually perform the
+upgrade there will be no need to contact that slow and/or unreliable registry
+server.
+
+#### Pre-load and pin custom images
+
+As an engineer managing a cluster I want to be able to pin and pre-load custom
+images required to upgrade my own applications.
+
+#### Prepare an upgrade bundle
+
+As an engineer managing a cluster I want to be able to assemble an upgrade
+bundle that contains all the artifacts (container images and metadata) needed
+to upgrade the OpenShift cluster and my own applications. I want to hand over
+this upgrade bundle and the documentation explaining how to use it to the
+technicians that will perform the upgrade in a suitable media, for example a
+USB stick.
+
+#### Include custom images in the upgrade bundle
+
+As an engineer managing a cluster I want to be able to include in the upgrade
+bundle the images required to upgrade my own workloads.
+
+#### Explicitly allow vetted upgrade bundle
+
+As an engineer managing a cluster I want be able to explicitly approve the use
+of an upgrade bundle, so that only the bundle that I tested and vetted will be
+applied to the cluster.
+
+#### Upgrade a single-node or multi-node clusters using a bundle
+
+As a technician with little or no OpenShift knowledge I want to be able to
+upgrade a single-node or multi-node cluster using the upgrade bundle and its
+documentation. I can't bring up any additional infrastructure at the cluster
+site, in particular I can't bring up an image registry server, neither outside
+of the cluster nor inside. I want to plug the USB stick provided by the
```

Comment:
Starting a registry server inside single-node clusters is unacceptable because it introduces a chicken-egg problem if the registry server is a workload managed by the cluster.

For multi-node clusters an image registry server supported by Red Hat inside the cluster is unacceptable (that means full Quay) because it requires a large amount of resources (specially storage, requires ODF) and additional subscriptions. We have confirmed that the proposed mechanism is better in these regards: it doesn't need additional disks, it doesn't require a full Quay instance and doesn't require storage or registry workloads running after the upgrade has completed.

Reply:
Quay supports a use case very similar to this: https://docs.openshift.com/container-platform/4.13/installing/disconnected_install/installing-mirroring-creating-registry.html

Reply:
Yes, but that is only supported in an external machine, not inside the cluster. In the context that we are trying to cover we can't ask the customer to bring up an additional machine for that mirror registry.

Reply:
But the cluster is HA and has protections against workloads becoming non HA. Would local storage operator with the registry configured to use file based storage not work? Even ephemeral storage seems like it could be made to work with an init container that mirrored content from its peer. The image to run the registry itself would need to be pinned.

Reply:
The local storage operator requires an additional disk, and we can't guarantee that will be available in this context. Single-node clusters in particular often have only one disk.

What you describe is very close to our bundle approach. We use ephemeral storage: a temporary directory created from the bundle contents. We mirror content from peers: copy the bundle from peers. The registry server code is kind of pinned: it is part of CVO (or MCO, that is an implementation detail). In addition it only consumes resources during the upgrade, after that it goes away.

Reply:
> But the cluster is HA...

The local cluster is not actually highly available, because we have to deliver [RFE-4482][1]'s :

> I have a multi-node cluster that could have a complete power outage, and I want it to recover smoothly from that kind of disruption, despite the lack of access to image registries, local or remote.

For example, [powering on the cluster after reaching the user-site][3].  I [still][2] think we want to call that out in the use-case section of this enhancement, because it's the only reason I see that makes the in-cluster registry an insufficient solution.

[1]: https://issues.redhat.com/browse/RFE-4482
[2]: https://github.com/openshift/enhancements/pull/1432#issuecomment-1712441185
[3]: https://github.com/openshift/appliance/blob/master/docs/user-guide.md#high-level-flow-overview